engine/src/distribution/primitives.cpp

```diff
@@ -263,10 +269,16 @@ void distributePartitions(Context * context, std::vector<NodeColumnView> & parti
 		}
 		BlazingTableView columns = nodeColumn.second;
 		auto destination_node = nodeColumn.first;
-		threads.push(([message_id, context_token, self_node, destination_node, columns](int thread_id) mutable {
+		auto send_function = threads.push(([message_id, context_token, self_node, destination_node, columns](int thread_id) mutable {
 			auto message = Factory::createColumnDataMessage(message_id, context_token, self_node, columns);
 			Client::send(destination_node, *message);
 		}));
+
+		try {
+			send_function.get();
```

Comment:
blocks execution when it should be able to schedule as many of these as it needs. this needs to be caught by some global context that can then STOP the whole graph from executing on EVERY node

Reply:
To be clear we dont have to stop execution on every node for now but if we are using a pool we MUST have some lace to store these futures to make sure they all have get run on them before the query executes but they cant prevent the kernel from scheduling multiple simultaneously